scylla-cql/src/frame/response/result.rs

```diff
@@ -373,6 +377,145 @@ impl Row {
+    #[inline]
+    pub fn rows_iter<'r, R: DeserializeRow<'r>>(
+        &'r self,
+    ) -> StdResult<TypedRowIterator<'r, R>, ParseError> {
+        let slice = FrameSlice::new(&self.raw_rows);
+        let raw = RowIterator::new(self.rows_count, &self.metadata.col_specs, slice);
+        TypedRowIterator::new(raw)
+    }
+
+    /// Converts the `RawRows` into `Rows` - a legacy, inefficient representation.
+    ///
+    /// Provided only to make migration to the new deserialization API
+    /// more convenient - this function will be deprecated and removed
+    /// in future releases.
+    pub fn into_legacy_rows(self) -> StdResult<Rows, ParseError> {
+        let rows = self.rows_iter::<Row>()?.collect::<StdResult<_, _>>()?;
+        Ok(Rows {
+            metadata: self.metadata,
+            rows_count: self.rows_count,
+            rows,
+            serialized_size: self.raw_rows.len(),
+        })
+    }
+}
+
+// Technically not an iterator because it returns items that borrow from it,
+// and the std Iterator interface does not allow for that.
+// TODO: Move to row.rs
+/// A _lending_ iterator over serialized rows.
+///
+/// This type is similar to `RowIterator`, but keeps ownership of the serialized
+/// result. Because it returns `ColumnIterator`s that need to borrow from it,
+/// it does not implement the `Iterator` trait (there is no type in the standard
+/// library to represent this concept yet).
+#[derive(Debug)]
+pub struct RawRowsLendingIterator {
+    metadata: ResultMetadata,
+    remaining: usize,
+    at: usize,
+    raw_rows: Bytes,
+}
+
+impl RawRowsLendingIterator {
+    /// Creates a new `RawRowsLendingIterator`, consuming given `RawRows`.
+    #[inline]
+    pub fn new(raw_rows: RawRows) -> Self {
+        Self {
+            metadata: raw_rows.metadata,
+            remaining: raw_rows.rows_count,
+            at: 0,
+            raw_rows: raw_rows.raw_rows,
+        }
+    }
+
+    /// Returns a `ColumnIterator` that represents the next row.
+    ///
+    /// Note: the `ColumnIterator` borrows from the `RawRowsLendingIterator`.
+    /// The column iterator must be consumed before the rows iterator can
+    /// continue.
+    #[inline]
+    pub fn next(&mut self) -> Option<StdResult<ColumnIterator, ParseError>> {
+        self.remaining = self.remaining.checked_sub(1)?;
+
+        let mut mem = &self.raw_rows[self.at..];
+
+        // Skip the row here, manually
+        for _ in 0..self.metadata.col_specs.len() {
+            if let Err(err) = types::read_bytes_opt(&mut mem) {
+                return Some(Err(err));
+            }
+        }
+
+        let slice = FrameSlice::new_subslice(&self.raw_rows[self.at..], &self.raw_rows);
+        let iter = ColumnIterator::new(&self.metadata.col_specs, slice);
+        self.at = self.raw_rows.len() - mem.len();
+        Some(Ok(iter))
+    }
+
+    #[inline]
+    pub fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.remaining))
+    }
```

Comment:
The iterator can produce less rows if it fails to parse. It shouldn't usually be the case, but there is a possibility. If we always returned `self.remaining` as a lower bound, this could be incorrect.

I'm not entirely sure though whether we should specify `self.remaining` or `0` as a lower bound. Although [the documentation](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.size_hint) specifies that this is just a hint for optimization purposes, it also states that returning wrong bounds is a "violation of the trait's protocol". I found some discussions which seem to agree with the interpretation that wrong bounds are a bug; the code that calls it can assume that this implementation is correct provided that it doesn't do anything unsafe under that assumption (but it's e.g. perfectly fine to panic) - for example, see here: https://internals.rust-lang.org/t/size-hint-correctness-reproducibility-and-documentation/8058 .

On the other hand, providing a tighter lower bound indeed could lead to better optimization, but I'd rather not risk it because it's not clear to me whether an implementation of `size_hint` that is sometimes incorrect can lead to panics in the caller.